src/Servers.jl

```diff
@@ -130,278 +109,196 @@ function check_rate_limit(tcp;
     end
 
     HTTP.listen() do request::HTTP.Request
         @show HTTP.header(request, "Content-Type")
-        @show HTTP.payload(request)
+        @show HTTP.payload(request, String)
         return HTTP.Response(404)
     end
 ```
 """
-listen(f, host::IPAddr=Sockets.localhost, port::Integer=8081; kw...) = listen(f, Sockets.InetAddr(host, port); kw...)
-listen(f, host::AbstractString, port::Integer; kw...) = listen(f, parse(IPAddr, host), port; kw...)
-listen(f, host::AbstractString; kw...) = listen(f, string(host); kw...)
+function listen end
+
+const nolimit = typemax(Int)
 
-function listen(f::Function,
-                host::Union{Sockets.InetAddr, String},
+getinet(host::String, port::Integer) = Sockets.InetAddr(parse(IPAddr, host), port)
+getinet(host::IPAddr, port::Integer) = Sockets.InetAddr(host, port)
+
+function listen(f,
+                host::Union{IPAddr, String}=Sockets.localhost,
+                port::Integer=8081
                 ;
-                ssl::Bool=false,
-                require_ssl_verification::Bool=true,
-                sslconfig::Union{SSLConfig, Nothing}=nothing,
-                pipeline_limit::Int=ConnectionPool.default_pipeline_limit,
-                tcpisvalid::Function=(tcp; kw...)->true,
-                tcpref::Ref=Ref{Base.IOServer}(),
+                stream::Bool=false,
+                sslconfig::Union{MbedTLS.SSLConfig, Nothing}=nothing,
+                tcpisvalid::Union{Function, Nothing}=nothing,
+                tcpref::Union{Ref, Nothing}=nothing,
                 reuseaddr::Bool=false,
-                connectioncounter::Base.RefValue{Int}=Ref(0),
-                kw...)
+                connectioncounter::Ref{Int}=Ref(0),
+                ratelimit::Union{Rational{Int}, Nothing}=nothing,
+                reuse_limit::Int=nolimit, readtimeout::Int=0,
+                verbose::Bool=false)
 
-    if ssl && sslconfig === nothing
-        sslconfig = SSLConfig(require_ssl_verification)
-    end
+    handler = f isa Handler ? f : stream ? StreamHandlerFunction(f) : RequestHandlerFunction(f)
 
-    @info "Listening on: $host"
-    if isassigned(tcpref)
+    inet = getinet(host, port)
+    if tcpref !== nothing
         tcpserver = tcpref[]
     elseif reuseaddr
         tcpserver = Sockets.TCPServer(; delay=false)
-        if Sys.islinux() || Sys.isapple()
+        if Sys.isunix()
             rc = ccall(:jl_tcp_reuseport, Int32, (Ptr{Cvoid},), tcpserver.handle)
-            Sockets.bind(tcpserver, host.host, host.port; reuseaddr=true)
+            Sockets.bind(tcpserver, inet.host, inet.port; reuseaddr=true)
         else
             @warn "reuseaddr=true may not be supported on this platform: $(Sys.KERNEL)"
-            Sockets.bind(tcpserver, host.host, host.port; reuseaddr=true)
+            Sockets.bind(tcpserver, inet.host, inet.port; reuseaddr=true)
         end
         Sockets.listen(tcpserver)
     else
-        tcpserver = Sockets.listen(host)
-        tcpref[] = tcpserver
```

Comment:
Need to keep this line (and duplicate it in the `elseif reuseaddr` block ?) otherwise the documented use of "e.g. to allow closing the server." won't work.

This code will probably fall out simpler if the seperate `server` parameter is added per suggestion on line 303 above.